sycl/test/basic_tests/macros_no_rdc.cpp

```diff
@@ -0,0 +1,22 @@
+// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple -fsycl-device-only -E -dD -fno-sycl-rdc %s -o %t.device
+// RUN: %clangxx %fsycl-host-only -fno-sycl-rdc -E -dD %s -o %t.host
+//
+// RUN: FileCheck --match-full-lines %s < %t.device --check-prefixes=DEVICE-FULL-LINE
+// RUN: FileCheck --match-full-lines %s < %t.host --check-prefixes=HOST
+//
+// Remove __DPCPP_SYCL_EXTERNAL to simplify regex for DEVICE prefix
+// RUN: sed -i 's|__DPCPP_SYCL_EXTERNAL||g' %t.device
+// RUN: FileCheck %s < %t.device --check-prefixes=DEVICE
+//
+// With -fno-sycl-rdc, device code should not define or use SYCL_EXTERNAL
+// DEVICE-FULL-LINE-NOT: #define SYCL_EXTERNAL
+// DEVICE-FULL-LINE-DAG: #define __DPCPP_SYCL_EXTERNAL __attribute__((sycl_device))
```

Comment:
Please note that: [`CHECK-NOT`](https://llvm.org/docs/CommandGuide/FileCheck.html#the-check-not-directive) is bound by other matches. I.e. this test check that there is no `#define SYCL_EXTERNAL` **prior** to `#define __DPCPP_SYCL_EXTERNAL`, but it won't catch `#define SYCL_EXTERNAL` later in the code. You should either add a second `CHECK-NOT` or use `--implicit-check-not` instead.

[`CHECK-DAG`](https://llvm.org/docs/CommandGuide/FileCheck.html#the-check-dag-directive) should be used when you have several checks which should happen in arbitrary order. However, it is still bound by surrounding directives other than `CHECK-DAG`, i.e. `CHECK-DAG: a`, then `CHECK-NOT: b`, then `CHECK-DAG: c` won't match `c a`. In your example, since there is only one directive, you can use a regular `CHECK`, because `CHECK-DAG` provides you with no benefits.

Reply:
thanks a lot for catching this, i should have fixed in it the commit i just pushed. i need to mess around with those a bit more, i don't have a good intuition for the semantics.